src/ports/postgres/modules/pred_metrics/pred_metrics.py_in

```diff
@@ -0,0 +1,391 @@
+    sql_st2= ""
+    sql_st3= """ FROM {table_in} """.format(**locals())
+    sql_st4= ""
+    if grouping_cols:
+        sql_st2= """ , {grouping_cols} """.format(**locals())
+        sql_st4= """ GROUP BY {grouping_cols}""".format(**locals())
+    sql_st = sql_st1+sql_st2+sql_st3+sql_st4
+    plpy.execute(sql_st)
+
+def mean_squared_error(
+	schema_madlib, table_in, table_out, prediction_col, observed_col,
+    grouping_cols=None):
+    sql_st1 = """
+        CREATE TABLE {table_out} AS
+        SELECT
+            SUM(({prediction_col}- {observed_col})^2)
+            /COUNT(*) AS mean_squared_error """.format(**locals())
+    sql_st2= ""
+    sql_st3= """ FROM {table_in} """.format(**locals())
+    sql_st4= ""
+    if grouping_cols:
+        sql_st2= """ , {grouping_cols} """.format(**locals())
+        sql_st4= """ GROUP BY {grouping_cols}""".format(**locals())
+    sql_st = sql_st1+sql_st2+sql_st3+sql_st4
+    plpy.execute(sql_st)
+
+
+def __r2_score(
+	schema_madlib, table_in, table_out, prediction_col, observed_col):
+
+    sql_st = """
+        SELECT
+            SUM({observed_col})/COUNT(*) AS mean
+        FROM {table_in}
+        """.format(**locals())
+    mean = plpy.execute(sql_st)[0]['mean']
+
+    sql_st = """
+        SELECT
+            SUM(({prediction_col}- {observed_col})^2) AS ssres
+        FROM {table_in}
+        """.format(**locals())
+    ssres = plpy.execute(sql_st)[0]['ssres']
+
+    sql_st = """
+        SELECT
+            SUM(({observed_col}- {mean})^2) AS sstot
+        FROM {table_in}
+        """.format(**locals())
+    sstot = plpy.execute(sql_st)[0]['sstot']
+
+    sql_st = """
+        CREATE TABLE {table_out} AS
+        SELECT 1-({ssres}/{sstot}) AS r2_score
+        """.format(**locals())
+    plpy.execute(sql_st)
+
+def r2_score(
+	schema_madlib, table_in, table_out, prediction_col, observed_col,
+    grouping_cols):
+
+    if grouping_cols is None:
+        __r2_score(schema_madlib, table_in, table_out, prediction_col
+            , observed_col)
+        return
+    ## temp table naming
+    mean_tmp_table = "pg_temp." + unique_string() + "_mean"
+    ssres_tmp_table = "pg_temp." + unique_string() + "_ssres"
+    sstot_tmp_table = "pg_temp." + unique_string() + "_sstot"
+    sql_st = """
+        CREATE TABLE {mean_tmp_table} AS
+        SELECT
+            SUM({observed_col})/COUNT(*) AS mean, {grouping_cols}
+        FROM {table_in}
+        GROUP BY {grouping_cols}
+        """.format(**locals())
+    plpy.execute(sql_st)
+
+    sql_st = """
+        CREATE TABLE {ssres_tmp_table} AS
+        SELECT
+            SUM(({prediction_col}- {observed_col})^2) AS ssres, {grouping_cols}
+        FROM {table_in}
+        GROUP BY {grouping_cols}
+        """.format(**locals())
+    plpy.execute(sql_st)
+
+    sql_st = """
+        CREATE TABLE {sstot_tmp_table} AS
+        SELECT
+            SUM(({observed_col}- mean)^2) AS sstot, table_in.{grouping_cols}
```

Comment:
I'm pretty sure this will blow up if {grouping_cols} actually has more than one column, unless you're expected to define them as something like "column_1, table_in.column_2, table_in.column_3", which is obviously very ugly. It would be better to see if grouping_cols is a sequence and if it is do the necessary map().